### ISSUE TITLE
LeechBlock doesn't get along well with Firefox's GC

### ISSUE THREAD

**Issue by benburrill** [proginosko/LeechBlockNG#124]
I've been having memory issues for a while now where Firefox's memory usage steadily rises (a couple MB/s), sometimes falling back on its own, but disturbingly often it just keeps on rising seemingly forever.  Disabling LeechBlock stops the memory from rising.

To deal with this, whenever my computer starts getting slow, I go to about:memory, clear the GC, and then turn swap on and off again.  This is annoying, but it's better for my productivity than disabling LeechBlock!

When measuring memory in about:memory, the problem shows up as `structured-clone-holder` and I think the problem is likely related to [this Firefox issue](https://bugzilla.mozilla.org/show_bug.cgi?id=1476032).  Specificially, I believe this problem is because LeechBlock sends messages to tabs every second to update the time remaining, creating those structured-clone-holders at such a frequency that Firefox's GC can't keep up.  However, I don't really see why the background script needs to be responsible for the countdown.  The tabs can maintain their own timer which can be updated by the background script whenever you switch to the tab, but for the most part the tabs can just do the count-down on their own.

I'm on Linux and I'm currently using Firefox 69.  I can try updating to FF 70 if you want, but I kinda doubt it will fix it.  If I'm right about the cause, even if FF does improve the way the GC works, I think it's likely LeechBlock will still perform better if it's not sending updates to the tabs every second.

**Comment by martinxyz** [comment]
Just to confirm, this is a problem for me, too.

With a couple of tabs open, I'm measuring a growth of ~7MB per minute of the "WebExtensions" process. Firefox is just sitting in the background. Disabling LeechBlockNG stops the leak. (Measured without restarting Firefox. Firefox 71 on Linux. With some other extensions also running.)

![image](https://user-images.githubusercontent.com/59041/71468283-ad6f1400-27c5-11ea-9cec-723d20b6648e.png)

**Comment by benburrill** [comment]
The workaround I've settled on is turning on the "Process only active tabs" setting.  This completely (or at least to the extent that I can notice) fixes the problem.  However, depending on your block settings, this may not be the behavior you want.  I somewhat recently changed my block settings to be entirely delay-based, and "Process only active tabs" works flawlessly with that since there is no need to update any other tabs.

Of course the issue still exists and needs to be fixed, but for my purposes LeechBlock thankfully works sensibly again.

**Comment by martinxyz** [comment]
Great workaround, thank you. I've also settled on delay-based blocking long ago. At first it felt like cheating to get around the block just by standing up and doing some quick other task. Until I realized that what a positive change this was :-)

**Comment by proginosko** [comment]
I wonder if this is a Linux-specific issue. I'm using FF on Win10 and I haven't encountered any LB memory-usage issues. Even with a lot of tabs open, LB usage remains constant at around 550KB.

**Comment by benburrill** [comment]
Yeah, I figure it is Linux-specific.  The Firefox issue I linked was also reported by a Linux user.  I have not tried to reproduce on another OS though.

**Comment by martinxyz** [comment]
More testing:
* I can reliably reproduce memory growth of 0.5MB/s with ~20 tabs. But in some test cases the growth is not permanent, e.g. with a fresh Firefox profile it jumps back every ~10 seconds.
* I can reproduce this with an empty LeechBlockNG config. (No blocking configured. Loaded as a temporary extension, git master 4075ff, with a new manifest-gecko-id.)

Code exploration:
* background.js never calls browser.tabs.sendMessage() while I'm viewing about:debugging, yet memory grows.
* Even with an empty config there is a lot of code running `for every second { for each tab { for all 6 sets { ... }}}`. Much is skipped by not passing testURL(), but still this seems excessive when Firefox is in the background. This polling is using about 0.5% of my CPU, probably much more if most tabs are in a block-set. Sadly the profiler doesn't show any details.
* IMO processing only the active tab(s) is a good-enough solution. I don't see the point of doing anything with hidden tabs until they become active. But I'm not familiar with all use-cases. My habit is to accumulate hidden tabs as a sort-of weekly reading list.

**Comment by rakyi** [comment]
I'm seeing this issue in Firefox 132.0.2 on macOS 15.1, but I observed it also in the past on older versions. I usually have at least a few dozen tabs open across 1–3 windows.

**Comment by Jorpho** [comment]
After dealing with out-of-memory crashes every few hours for months on end, I finally decided to disable Leachblock and everything is finally ticking along smoothly again. (I thought moving from Windows 10 to Windows 11 might somehow fix it, but it did not.) 

I also have a few dozen tabs running in a single mostly-private window. I had an inkling Leachblock might be responsible when I checked in on about:memory on occasion, but the crashes were so unpredictable that I could never be sure.

(Crashing the browser is definitely one way to push me to stop getting distracted and get back to work.)

Leechblock 1.7, Firefox 143 32-bit.

**Comment by martinxyz** [comment]
Did the workaround (mentioned above) help?

Options -> General -> (Show Full Options) -> Process only active tabs

**Comment by Jorpho** [comment]
Looks promising so far. Rather makes me wish I'd found this thread months ago. And makes me wonder how many other people might be running into this bug without realizing it.
